### PR TITLE
Assignment5

### PR DESCRIPTION
**Pull request by jschofield1**
Collaborated with @coproduct

    Create a class library project called "GenericsHomework.". ✔
    Create a node class that can contain a value of any type and points to the next node and traversing the next node points back to the first item.
        Define the Node class
        Include a constuctor that takes a value. (No validation is necessary on the value). ✔
        Add a ToString() override that writes out the value's ToString() result. ✔
        Add a Next property that references the next node or else refers back to itself if there are no other nodes in the list. ✔
            The Next property should be non-nullable (careful to follow the non-nullable property guidelines) ✔
            The Next property setter should be private. ✔
            In addition to non-null validation, the body of the Next property should insert the next Node (the value) into the list. ✔
        Add an Insert method that takes a value and inserts a new Node instance after the current node (by invoking the Next property). ✔
        Add a clear method that effectively removes all items from a list except the current node. Pay attention as to whether you should be concerned with the following:
            Whether it is sufficient to only set Next to itself ✔
            Whether to set the removed items to circle back on themselves. In other words, whether to close the loop of the removed items. (Provide a test to show why this is required if it is required). ✔
            Given there is a circular list of items, provide a comment to indicate whether you need to worry about garbage colleciton because all the items point to each other and therefore may never be garbage collected. ✔
    You should not rely on any BCL generic classes for your implementation. ✔

Extra Credit

Do one of the following two options (or both if you want extra extra credit) :)

    Implement a VennDiagram structure that contains n Circles that only contains homogenous reference types of any type. ❌

    Each circle contains n items and each item can belong to one more more Circle instances.
    You are not required to use a Node from earlier in the homework for your venn diagram implementation.
    You are welcome to use exising BCL generic classes for the extra credit.

    Implement Systm.Collections.Generic.ICollection<T> on the Node class ✔

Fundamentals

    Ensure you enable nullable reference types, net5 targetted, C# 9.0, warnings as errors, and enabled .NET analyzers for both projects ✔
        For this assignment, always use Assert.AreEquals<T>() (the generic version)
    All of the above should be unit tested ✔

### CODE REVIEW

**Comment by MeguminXIII** [comment]
I will review

**Review by MeguminXIII — COMMENTED** [review]
Create a class library project called "GenericsHomework.". ✔
Create a node class that can contain a value of any type and points to the next node and traversing the next node points back to the first item.
Define the Node class
Include a constuctor that takes a value. (No validation is necessary on the value). ✔
Add a ToString() override that writes out the value's ToString() result. ✔
Add a Next property that references the next node or else refers back to itself if there are no other nodes in the list. ✔
The Next property should be non-nullable (careful to follow the non-nullable property guidelines) ✔
The Next property setter should be private. ❌
setter is not set to private, an example: public Property prop { get; private set; }

In addition to non-null validation, the body of the Next property should insert the next Node (the value) into the list. ✔
Add an Insert method that takes a value and inserts a new Node instance after the current node (by invoking the Next property). ✔
Add a clear method that effectively removes all items from a list except the current node. Pay attention as to whether you should be concerned with the following:
Whether it is sufficient to only set Next to itself ✔
Whether to set the removed items to circle back on themselves. In other words, whether to close the loop of the removed items. (Provide a test to show why this is required if it is required). ❌
Did not do this option.

Given there is a circular list of items, provide a comment to indicate whether you need to worry about garbage collection because all the items point to each other and therefore may never be garbage collected. ❌
No comment in or around Clear() talking about garbage collection

You should not rely on any BCL generic classes for your implementation. ✔

Extra Credit
Do one of the following two options (or both if you want extra extra credit) :)

Implement a VennDiagram structure that contains n Circles that only contains homogenous reference types of any type. ❌
Each circle contains n items and each item can belong to one more more Circle instances.
You are not required to use a Node from earlier in the homework for your venn diagram implementation.
You are welcome to use exising BCL generic classes for the extra credit.
Implement Systm.Collections.Generic.ICollection<T> on the Node class ✔
Fundamentals
Ensure you enable nullable reference types, net5 targetted, C# 9.0, warnings as errors, and enabled .NET analyzers for both projects ✔
For this assignment, always use Assert.AreEquals<T>() (the generic version)❌
Did not use generic AreEquals in:
GetArrayLength_CreateMultipleNode_ArrayEqualsExpectedLength()
CopyTo_CopyValuesToArray_ArrayHoldsExpectedValues()
Remove_RemoveSpecificValue_ExpectedValueRemoved()
GetEnumerator_ForeachToIterate_AddsExpectedValues()

All of the above should be unit tested ✔

Good Job

**Comment by jschofield1** [comment]
Create a class library project called "GenericsHomework.". ✔
    Create a node class that can contain a value of any type and points to the next node and traversing the next node points back to the first item.
        Define the Node class
        Include a constuctor that takes a value. (No validation is necessary on the value). ✔
        Add a ToString() override that writes out the value's ToString() result. ✔
        Add a Next property that references the next node or else refers back to itself if there are no other nodes in the list. ✔
            The Next property should be non-nullable (careful to follow the non-nullable property guidelines) ✔
            The Next property setter should be private. ✔
            In addition to non-null validation, the body of the Next property should insert the next Node (the value) into the list. ✔
        Add an Insert method that takes a value and inserts a new Node instance after the current node (by invoking the Next property). ✔
        Add a clear method that effectively removes all items from a list except the current node. Pay attention as to whether you should be concerned with the following:
            Whether it is sufficient to only set Next to itself ✔
            Whether to set the removed items to circle back on themselves. In other words, whether to close the loop of the removed items. (Provide a test to show why this is required if it is required). ✔
            Given there is a circular list of items, provide a comment to indicate whether you need to worry about garbage colleciton because all the items point to each other and therefore may never be garbage collected. ✔
    You should not rely on any BCL generic classes for your implementation. ✔

Extra Credit

Do one of the following two options (or both if you want extra extra credit) :)

    Implement a VennDiagram structure that contains n Circles that only contains homogenous reference types of any type. ❌

    Each circle contains n items and each item can belong to one more more Circle instances.
    You are not required to use a Node from earlier in the homework for your venn diagram implementation.
    You are welcome to use exising BCL generic classes for the extra credit.

    Implement Systm.Collections.Generic.ICollection<T> on the Node class ✔

Fundamentals

    Ensure you enable nullable reference types, net5 targetted, C# 9.0, warnings as errors, and enabled .NET analyzers for both projects ✔
        For this assignment, always use Assert.AreEquals<T>() (the generic version)
    All of the above should be unit tested ✔

File: Assignment5/GenericsHomework/GenericsHomework/Node.cs [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+            get
+            {
+                return _Data!;
+            }
+            private set
+            {
+                _Data = value ??
+                    throw new ArgumentNullException(nameof(value));
+            }
+        }
```
Comment (mmwoodfo) [inline Assignment5/GenericsHomework/GenericsHomework/Node.cs:22]: You don't need to do validation on `_Data` here. Just keep the private variable and remove the public property.

File: Assignment5/GenericsHomework/GenericsHomework/Node.cs [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+        private Node<T>? _Next;
+
+        public Node<T> Next
+        {
+            get
+            {
+                return _Next!;
+            }
+            private set
```
Comment (mmwoodfo) [inline Assignment5/GenericsHomework/GenericsHomework/Node.cs:32]: Submitted late in [this](https://github.com/IntelliTect-Samples/EWU-CSCD371-2021-Winter/pull/61/commits/547237a1d0eada310b237041867c824fe8289760) PR.

File: Assignment5/GenericsHomework/GenericsHomework/Node.cs [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+            return Data.ToString()!;
+        }
+
+        public void Insert(T value)
+        {
+            Node<T> newNode = new(value);
+            Next = newNode;
+        }
+
+        /* No need to worry about garbage collection in the case of a circular list, garbage collector handles circular references */
```
Comment (mmwoodfo) [inline Assignment5/GenericsHomework/GenericsHomework/Node.cs:86]: Submitted late in [this](https://github.com/IntelliTect-Samples/EWU-CSCD371-2021-Winter/pull/61/commits/547237a1d0eada310b237041867c824fe8289760) PR.

File: Assignment5/GenericsHomework/GenericsHomework/Node.cs [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+                return _Data!;
+            }
+            private set
+            {
+                _Data = value ??
+                    throw new ArgumentNullException(nameof(value));
+            }
+        }
+
+        private Node<T>? _Next;
```
Comment (mmwoodfo) [inline Assignment5/GenericsHomework/GenericsHomework/Node.cs:24]:
```suggestion
        private Node<T> _Next;
```
If you set this in the constructor it does not need to be nullable.

File: Assignment5/GenericsHomework/GenericsHomework/Node.cs [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+        public bool IsReadOnly
+        {
+            get;
+        }
+
+        public Node(T value)
+        {
+            Data = value;
+            Next = this;
```
Comment (mmwoodfo) [inline Assignment5/GenericsHomework/GenericsHomework/Node.cs:69]:
```suggestion
            _Data = value;
            _Next = this;
```

File: Assignment5/GenericsHomework/GenericsHomework/Node.cs [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+        }
+
+        public override string ToString()
+        {
+            if (Data is null) 
+                throw new ArgumentNullException(nameof(Data));
```
Comment (mmwoodfo) [inline Assignment5/GenericsHomework/GenericsHomework/Node.cs:75]:
```suggestion
            if (Data is null) 
                throw new ArgumentNullException(nameof(Data));
```
Consider returning an empty string if `Data` is null.

File: Assignment5/GenericsHomework/GenericsHomework/Node.cs [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+
+            return Data.ToString()!;
```
Comment (mmwoodfo) [inline Assignment5/GenericsHomework/GenericsHomework/Node.cs:77]:
```suggestion
            return Data?.ToString() ?? "";
```
This will return an empty string when null. Avoid using `!` if you can help it.

File: Assignment5/GenericsHomework/GenericsHomework.Tests/NodeTests.cs [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+            Node<int> first = new Node<int>(intNum);
+            Node<double> second = new Node<double>(dblNum);
+            Node<string> third = new Node<string>(strNum);
+            Node<char> fourth = new Node<char>(chrNum);
+
+            Assert.AreEqual<string>(intNum.ToString(), first.ToString());
+            Assert.AreEqual<string>(dblNum.ToString(), second.ToString());
+            Assert.AreEqual<string>(strNum.ToString(), third.ToString());
+            Assert.AreEqual<string>(chrNum.ToString(), fourth.ToString());
```
Comment (mmwoodfo) [inline Assignment5/GenericsHomework/GenericsHomework.Tests/NodeTests.cs:84]:
```suggestion
            Assert.AreEqual(intNum.ToString(), first.ToString());
            Assert.AreEqual(dblNum.ToString(), second.ToString());
            Assert.AreEqual(strNum.ToString(), third.ToString());
            Assert.AreEqual(chrNum.ToString(), fourth.ToString());
```

File: Assignment5/GenericsHomework/GenericsHomework.Tests/NodeTests.cs [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+        [TestMethod]
+        public void Insert_InsertNewNodeOnHead_NextOnLastNodeLoopsToHead()
+        {
+            Node<int> node = new Node<int>(1);
+            
+            node.Insert(2);
+
+            Assert.AreEqual<int>(1, node.Data);
+            Assert.AreEqual<int>(2, node.Next.Data);
+            Assert.AreEqual<int>(1, node.Next.Next.Data);
```
Comment (mmwoodfo) [inline Assignment5/GenericsHomework/GenericsHomework.Tests/NodeTests.cs:96]: You should test more than two nodes being inserted into the list. You should also test the order in which the nodes get added to the list. When writing unit tests the goal is to write unit tests that stress test your code. Hit the limits, test edge cases, and hopefully reveal bugs.

The following unit test will cause a failure in your code.
```C#
// 1 -> 2 -> 3
// ┖────┘

// Arrange
Node<int> one = new Node<int>(1);

// Act
one.Next.Insert(2);
var two = one.Next;

two.Next.Insert(3);
var three = two.Next;

// Assert (These will fail with your code)
Assert.AreEqual(one, two.Next.Next);
Assert.AreEqual(one.Next.Next, three);
Assert.AreEqual(three, two.Next);
```

Instead of a linked list that looks like this:
```
1 -> 2 -> 3
┖─────────┘
```

You'll have this:
```
1 -> 3
┖────┘

2 -> 1 -> 3
     ┖────┘
```

I see you made a comment about this yesterday - so be sure to add robust tests in the future.

There are easy ways within visual studio to look at your variables and data structures. [Here are some tips for that](https://devblogs.microsoft.com/devops/7-ways-to-look-at-the-values-of-variables-while-debugging-in-visual-studio/).

File: Assignment5/GenericsHomework/GenericsHomework.Tests/NodeTests.cs [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+            int[] copiedArray = new int[3];
+            int[] expectedArray = { 1, 2, 3 };
+
+            Node<int> node = new Node<int>(1);
+
+            node.Insert(2);
+            node.Insert(3);
+            node.CopyTo(copiedArray, 0);
+
+            Assert.AreEqual<string>($"{expectedArray.ToString()}", $"{copiedArray.ToString()}");
```
Comment (mmwoodfo) [inline Assignment5/GenericsHomework/GenericsHomework.Tests/NodeTests.cs:132]: This is not doing what you think it's doing. The output of this test is: `Assert.AreEqual("System.Int32[]", "System.Int32[]");`. Which is true. Hence why your test passes.

You should use `Assert.IsTrue(expectedArray.SequenceEqual(copiedArray));` instead. This will fail, because of the comment on line 96 above.

File: Assignment5/GenericsHomework/GenericsHomework.Tests/NodeTests.cs [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+            
+            node.Insert(2);
+
+            Assert.AreEqual<int>(1, node.Data);
+            Assert.AreEqual<int>(2, node.Next.Data);
+            Assert.AreEqual<int>(1, node.Next.Next.Data);
+        }
+
+        [TestMethod]
+        public void Clear_ClearsArrayOnHead_OnlyHeadRemains()
```
Comment (mmwoodfo) [inline Assignment5/GenericsHomework/GenericsHomework.Tests/NodeTests.cs:100]: It would be good to have a unit test that tests if the garbage collector ever cleans up the detached nodes.

File: Assignment5/GenericsHomework/GenericsHomework.Tests/NodeTests.cs [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+            Assert.AreEqual<Node<int>>(node, nextNode);
+        }
+
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentNullException))]
+        public void ToString_PassesNullNode_ThrowsArgumentNullException()
+        {
+            Node<string> node = new Node<string>(null!);
+
+            node.ToString();
```
Comment (mmwoodfo) [inline Assignment5/GenericsHomework/GenericsHomework.Tests/NodeTests.cs:65]: This is not the line throwing the exception. This unit test is misleading in what it's actually testing. Your exception gets thrown on line 19 of your `Node` class when you try to set a null value to your `Data` property.

**Comment by mmwoodfo** [comment]
@jschofield1 I can't find a code review from you for Assignment5.

**Comment by jschofield1** [comment]
> 
> 
> @jschofield1 I can't find a code review from you for Assignment5.

We were told on the first day of class that if we paired with a classmate we were not required to do peer reviews. My review of my own code is on the initial comment.

**Comment by mmwoodfo** [comment]
> > @jschofield1 I can't find a code review from you for Assignment5.
> 
> We were told on the first day of class that if we paired with a classmate we were not required to do peer reviews. My review of my own code is on the initial comment.

@jschofield1 No problem. I missed that you mentioned in your comment you paired. I'll update both your score and @coproduct's score accordingly. Note that this may take a few days. Thank you!
